cpp/src/arrow/compare.cc

```diff
@@ -459,6 +473,26 @@ class RangeDataEqualsImpl {
     return Status::OK();
   }
 
+  template <typename RunEndsType>
+  Status CompareRunEndEncoded() {
+    auto left_span = ArraySpan(left_);
+    auto right_span = ArraySpan(right_);
+    left_span.SetSlice(left_.offset + left_start_idx_, range_length_);
+    right_span.SetSlice(right_.offset + right_start_idx_, range_length_);
+    for (auto it = ree_util::MergedRunsIterator<RunEndsType, RunEndsType>(left_span,
+                                                                          right_span);
```

Comment:
That only works if both run-ends types are identical, which is inconsistent with type equality definition below.